### PR TITLE
DLPack overhaul

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**New feature** (*non-breaking change which adds functionality*)
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
This PR reworks the DLPack support in DALI.
New features:
- shared buffer ownership (not just views)
- add pinned memory support
Refactoring:
- remove DLTensorResource inheritance
- unify JAX DLTensorResource and other resources

### Tests:
Existing tests: ExternalSource, PythonFunction, JAX, dali_test.bin:*DL*
New tests: buffer sharing (reference counting test).
- [X] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/data/dltensor.h [hunk 1]
```diff
@@ -25,55 +28,275 @@
...
+ */
+inline std::string to_string(const DLDataType &dl_type) {
+  const char *code_str[] = {
+    "i", "u", "f", "p", "bf", "c", "b"
+  };
+  std::stringstream ss;
+  if (dl_type.code < std::size(code_str))
+    ss << code_str[dl_type.code];
+  else
+    ss << "<unknown" << dl_type.code + 0 << ">";
```
Comment (banasraf) [inline dali/pipeline/data/dltensor.h:71]:
```suggestion
    ss << "<unknown:" << dl_type.code + 0 << ">";
```

File: dali/pipeline/data/dltensor.h [hunk 2]
```diff
@@ -25,55 +28,275 @@
...
+DLL_PUBLIC DLDataType ToDLType(DALIDataType type);
+
+/** Converts a DLPack type to DALI type. */
+DLL_PUBLIC DALIDataType ToDALIType(const DLDataType &dl_type);
+
+/** Returns type string for given DLPack type
+ *
+ * The text representation looks like:
+ * <type><bits>[x<lanes>]
+ * with <lanes>x present only if the number of lanes is > 1
```
Comment (banasraf) [inline dali/pipeline/data/dltensor.h:49]:
```suggestion
 * with x<lanes> present only if the number of lanes is > 1
```

File: dali/pipeline/data/dltensor.h [hunk 3]
```diff
@@ -25,55 +28,275 @@
...
+ * |   +- *deleter            |  |
+ * |                          |  |
+ * +-- Payload                |  |
+ *     |                      |  |
+ *     +-- shape  <-----------+  |
+ *     +-- strides  <------------+
+ *     +-- ...
+ * ```
+ *
+ * You can use any payload structure of your choice, but it must privde the storage for DLTensor's
```
Comment (banasraf) [inline dali/pipeline/data/dltensor.h:136]:
```suggestion
 * You can use any payload structure of your choice, but it must provide the storage for DLTensor's
```

**Review by banasraf — APPROVED** [review]

File: dali/pipeline/data/dltensor.h [hunk 4]
```diff
@@ -25,55 +28,275 @@
...
+template <typename Backend>
+std::vector<DLMTensorPtr> GetSharedDLTensorList(TensorList<Backend> &tensor_list) {
+  int device_id = tensor_list.device_id();
+  bool pinned = tensor_list.is_pinned();
+
+  std::vector<DLMTensorPtr> dl_tensors{};
+  dl_tensors.reserve(tensor_list.num_samples());
+
+  for (int i = 0; i < tensor_list.num_samples(); ++i)
+    dl_tensors.push_back(GetDLTensorView(tensor_list[i], pinned, device_id));
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/data/dltensor.h:296]: Nice catch.

File: dali/pipeline/data/dltensor.cc [hunk 5]
```diff
@@ -19,18 +19,18 @@
 
 namespace dali {
 
-DLDataType GetDLType(DALIDataType type) {
+DLDataType ToDLType(DALIDataType type) {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/data/dltensor.cc:22]: Perhaps, but then again it's almost certainly faster than using TypeTable.

**Review by klecki — APPROVED** [review]

**Pull request merged**